src/EditorFeatures/Core/InheritanceMargin/AbstractInheritanceMarginService.cs

```diff
@@ -0,0 +1,159 @@
+
+            // Filter out System.Object. (otherwise margin would be shown for all classes)
+            var baseTypes = allBaseTypes
+                .WhereAsArray(type =>
+                    !(type is ITypeSymbol { SpecialType: SpecialType.System_Object }));
+
+            // Get all derived types
+            var derivedTypes = await GetDerivedTypesAndImplementationsAsync(
+                solution,
+                memberSymbol,
+                cancellationToken).ConfigureAwait(false);
+
+            if (!(baseTypes.IsEmpty && derivedTypes.IsEmpty))
+            {
+                var lineNumber = sourceText.Lines.GetLineFromPosition(declarationNode.SpanStart).LineNumber;
+                var item = await CreateInheritanceMemberInfoAsync(
+                    solution,
+                    memberSymbol,
+                    lineNumber,
+                    baseSymbols: baseTypes,
+                    derivedTypesSymbols: derivedTypes.OfType<ISymbol>().ToImmutableArray(),
+                    cancellationToken).ConfigureAwait(false);
+                builder.Add(item);
+            }
+        }
+
+        private static async Task AddInheritanceMemberItemsForTypeMembersAsync(
+            Solution solution,
+            SourceText sourceText,
+            SyntaxNode declarationNode,
+            ISymbol memberSymbol,
+            ArrayBuilder<InheritanceMemberItem> builder,
+            CancellationToken cancellationToken)
+        {
+            // Go up the inheritance chain to find all the overrides targets.
+            var overridenMembers = await SymbolFinder.FindOverridesArrayAsync(memberSymbol, solution, cancellationToken: cancellationToken).ConfigureAwait(false);
+
+            // Go up the inheritance chain to find all the implementing targets.
+            var implementingMembers = memberSymbol.ExplicitOrImplicitInterfaceImplementations();
+
+            // Go down the inheritance chain to find all overrides targets
+            var overridingMembers = GetOverridingSymbols(memberSymbol);
+
+            // Go down the inheritance chain to find all the implmeneing targets.
+            var implementedMembers = await GetImplementedSymbolsAsync(solution, memberSymbol, cancellationToken).ConfigureAwait(false);
+
+            if (!(overridenMembers.IsEmpty
+                && !overridingMembers.IsEmpty
+                && !implementingMembers.IsEmpty
+                && !implementedMembers.IsEmpty))
+            {
+                var lineNumber = sourceText.Lines.GetLineFromPosition(declarationNode.SpanStart).LineNumber;
+                var item = await CreateInheritanceMemberInfoForMemberAsync(
+                    solution,
+                    memberSymbol,
+                    lineNumber,
+                    implementingMembers: implementingMembers,
+                    implementedMembers: implementedMembers,
+                    overridenMembers: overridenMembers,
+                    overridingMembers: overridingMembers,
```

Comment:
does the inheritance item actually hold onto the symbols?  that's potentially a problem (keeping large graphs alive).  Should we hold onto symbokeys instead?

Reply:
No, it doesn't hold the symbol, it just holds line information, display name and DefinitionItem (for navigation purpose)

Reply:
👍